pkg/virtctl/ssh/native.go

```diff
@@ -0,0 +1,169 @@
+package ssh
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+
+	"github.com/golang/glog"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+	"golang.org/x/term"
+)
+
+func (o *SSH) prepareSSHClient(kind, namespace, name string) (*ssh.Client, error) {
+	streamer, err := o.prepareSSHTunnel(kind, namespace, name)
+	if err != nil {
+		return nil, err
+	}
+
+	conn := streamer.AsConn()
+	addr := fmt.Sprintf("%s/%s.%s", kind, name, namespace)
+	authMethods, err := o.getAuthMethods(kind, namespace, name)
+	if err != nil {
+		return nil, err
+	}
+
+	sshConn, chans, reqs, err := ssh.NewClientConn(conn,
+		addr,
+		&ssh.ClientConfig{
+			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+			Auth:            authMethods,
+			User:            sshUsername,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return ssh.NewClient(sshConn, chans, reqs), nil
+}
+
+func (o *SSH) getAuthMethods(kind, namespace, name string) ([]ssh.AuthMethod, error) {
+	var (
+		methods []ssh.AuthMethod
+		err     error
+	)
+
+	if len(identityFilePath) < 1 {
+		methods = trySSHAgent(methods)
+	} else {
+		methods, err = usePrivateKey(methods)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	methods = append(methods, ssh.PasswordCallback(func() (secret string, err error) {
+		password, err := readPassword(fmt.Sprintf("%s@%s/%s.%s's password: ", sshUsername, kind, name, namespace))
+		fmt.Println()
+		return string(password), err
+	}))
+
+	return methods, nil
+}
+
+func trySSHAgent(methods []ssh.AuthMethod) []ssh.AuthMethod {
+	socket := os.Getenv("SSH_AUTH_SOCK")
+	if len(socket) < 1 {
+		return methods
+	}
+	conn, err := net.Dial("unix", socket)
+	if err != nil {
+		glog.Error(err)
+		return methods
+	}
+	agentClient := agent.NewClient(conn)
+
+	return append(methods, ssh.PublicKeysCallback(agentClient.Signers))
+}
+
+func usePrivateKey(methods []ssh.AuthMethod) ([]ssh.AuthMethod, error) {
```

Comment:
Would it make sense to wrap the key fetching with `ssh.PublicKeysCallback` too? I just ask because I think we should only ask for the passphrase if we are going to try the public key. If the ssh agent already works we should not ask for the password.

Reply:
Even when using `ssh.PublicKeysCallback` if a publicKey with password is provided, it asks for a password. No checking if the key is actually needed.
Not sure how to workaround this, so if we default to `id_rsa` and there is a password on it, it will always ask for it :-/